packages/eslint-plugin/src/rules/prefer-string-starts-ends-with.ts

```diff
@@ -0,0 +1,652 @@
+          !isNumber(parentNode.right, 0) ||
+          !isStringType(node.object)
+        ) {
+          return;
+        }
+
+        context.report({
+          node: parentNode,
+          messageId: 'preferStartsWith',
+          fix(fixer) {
+            return fixWithArgument(
+              fixer,
+              parentNode,
+              'start',
+              parentNode.operator.startsWith('!'),
+            );
+          },
+        });
+      },
+
+      // foo.lastIndexOf('bar') === foo.length - 3
+      // foo.lastIndexOf(bar) === foo.length - bar.length
+      'BinaryExpression > CallExpression.left > MemberExpression.callee[property.name="lastIndexOf"][computed=false]'(
+        node: TSESTree.MemberExpression,
+      ): void {
+        const callNode = node.parent! as TSESTree.CallExpression;
+        const parentNode = callNode.parent!;
+
+        if (
+          callNode.arguments.length !== 1 ||
+          !isEqualityComparison(parentNode) ||
+          parentNode.left !== callNode ||
+          parentNode.right.type !== 'BinaryExpression' ||
+          parentNode.right.operator !== '-' ||
+          !isLengthExpression(parentNode.right.left, node.object) ||
+          !isLengthExpression(parentNode.right.right, callNode.arguments[0]) ||
+          !isStringType(node.object)
+        ) {
+          return;
+        }
+
+        context.report({
+          node: parentNode,
+          messageId: 'preferEndsWith',
+          fix(fixer) {
+            return fixWithArgument(
+              fixer,
+              parentNode,
+              'end',
+              parentNode.operator.startsWith('!'),
+            );
+          },
+        });
+      },
+
+      // foo.match(/^bar/) === null
+      // foo.match(/bar$/) === null
+      'BinaryExpression > CallExpression.left > MemberExpression.callee[property.name="match"][computed=false]'(
+        node: TSESTree.MemberExpression,
+      ): void {
+        const callNode = node.parent as TSESTree.CallExpression;
+        const parentNode = callNode.parent as TSESTree.BinaryExpression;
+        if (
+          !isEqualityComparison(parentNode) ||
+          !isNull(parentNode.right) ||
+          !isStringType(node.object)
+        ) {
+          return;
+        }
+
+        const parsed =
+          callNode.arguments.length === 1
+            ? parseRegExp(callNode.arguments[0])
+            : null;
+        if (parsed == null) {
+          return;
+        }
+
+        const { isStartsWith, text } = parsed;
+        context.report({
+          node: callNode,
+          messageId: isStartsWith ? 'preferStartsWith' : 'preferEndsWith',
+          *fix(fixer) {
+            if (!parentNode.operator.startsWith('!')) {
+              yield fixer.insertTextBefore(parentNode, '!');
+            }
+            yield fixer.replaceTextRange(
+              getPropertyRange(node),
+              `.${isStartsWith ? 'start' : 'end'}sWith`,
+            );
+            yield fixer.replaceText(
```

Comment:
This might not work for things like:
```typescript
const text = "212";
const search = 2;
text.match(search);
```

I am working on "prefer-regexp-exec" and copied some of your great stuff, but currently I am not sure how to implement a fixer.
These are my tests currently: https://github.com/ldrick/typescript-eslint/blob/5ed44d23c3278c15139e8da49e539c4e0b226a03/packages/eslint-plugin/tests/rules/prefer-regexp-exec.test.ts

Reply:
do we want fixer here? its sometimes better to not have it and let user decide how to fix stuff.
@bradzacher what do you think:

Best practices for fixes:
https://eslint.org/docs/developer-guide/working-with-rules#applying-fixes
> - **Avoid any fixes that could change the runtime behavior of code and cause it to stop working.**
> - Make fixes as small as possible. Fixes that are unnecessarily large could conflict with other fixes, and prevent them from being applied.
> - Only make one fix per message. This is enforced because you must return the result of the fixer operation from fix().
> - Since all rules are run again after the initial round of fixes is applied, it’s not necessary for a rule to check whether the code style of a fix will cause errors to be reported by another rule.

Reply:
My life would be easier 👍